designs/2022-languages/README.md

```diff
@@ -0,0 +1,689 @@
+
+The `esquery` [External class resolve](https://github.com/estools/esquery/pull/140) pull request implements an option that allows us to pass in a function to interpret pseudoclasses from `ESLintLanguage#matchesSelectorClass()`.
+
+#### The `parse()` Method
+
+The `parse()` method receives the file information and information about the ESLint context in order to correctly parse the file. This is a raw parse, meaning that `parse()` should do nothing more than convert the file into an abstract format without doing any postprocessing. This is important so that ESLint can tell how long parsing takes vs. other operations. (One problem with the current `parseForESLint()` approach is that we lose visibility into the true parsing time.)
+
+The `parse()` method is expected to return an object with at least `ast` and `body` properties; the object may also have other properties as necessary for the language to properly create a `SourceCode` object later. The return value from `parse()` is passed directly into `createSourceCode()`.
+
+This method will be called [inside of `Linter`](https://github.com/eslint/eslint/blob/28d190264017dbaa29f2ab218f73b623143cd1af/lib/linter/linter.js#L805) and will require that the language object be passed into the private `parse()` function.
+
+**Async:** For use with current ESLint, `parse()` must be synchronous; ESLint X will allow this method to be asynchronous.
+
+#### The `createSourceCode()` Method
+
+The `createSourceCode()` method is intended to create an instance of `SourceCode` suitable for ESLint and rule developers to use to work with the file. As such, it may perform postprocessing of a parse result in order to get it into a format that is easier to use.
+
+This method will be called [inside of `Linter`](https://github.com/eslint/eslint/blob/28d190264017dbaa29f2ab218f73b623143cd1af/lib/linter/linter.js#L828) and will require that the language object be passed into the private `parse()` function.
+
+**Async:** For use with current ESLint, `createSourceCode()` must be synchronous; ESLint X will allow this method to be asynchronous.
+
+### The `SourceCode` Object
+
+ESLint already has a `SourceCode` class that is used to [postprocess and organize the AST data](https://github.com/eslint/eslint/blob/f89403553b31d24f4fc841424cc7dcb8c3ef689f/lib/source-code/source-code.js#L149) for JavaScript files. This proposal extends the intended purpose of `SourceCode` such that it encapsulates all functionality related to a single file, including traversal. As such, `SourceCode` objects must implement the following interface:
+
+```ts
+interface SourceCode {
```

Comment:
I think this object should also have a `visitorKeys` property provided by the language/parser, so that we could [pass it to `esquery`](https://github.com/eslint/eslint/blob/b98fdd413a3b07b262bfce6f704c1c1bb8582770/lib/linter/linter.js#L1145-L1146) to properly work for that language.

Reply:
It seems like this should be on the `ESLintLanguage` object instead, as it's not unique to an individual file?

Reply:
That was my thinking as well, and I was actually about to leave the above comment on the `ESLintLanguage` interface instead of here.

But, if the JavaScript language plugin that we'll maintain will support custom parsers, then we'll need to take visitor keys from the parser that was used to parse the file rather than always using fixed visitor keys (`eslint-visitor-keys`) from the language object.

Reply:
For example, with `parser: "@typescript-eslint/parser"`, some selectors won't work well if we don't pass typescript-eslint's visitor keys to `esquery`:

https://github.com/eslint/eslint/issues/13639

Reply:
Ah good point. `SourceCode` it is.

Reply:
I added `visitorKeys` on both `ESLintLanguage` and `SourceCode`. I think the JS language is the exception case here, so we can have an optional `SourceCode#visitorKeys` that will override `ESLintLanguage#visitorKeys` when necessary.

Reply:
Sounds good to me!